### PR TITLE
Bitcoin support via SPV

### PR DESCRIPTION
**Pull request by Bushstar**
Add the ability to send and receive Bitcoins directly in the DeFi wallet via SPV. This PR adds the following commands.

spv_dumpprivkey ADDRESS
Return the private key that relates to the address provided.

spv_getbalance
Displays the Bitcoin balance.

spv_getnewaddress
Returns a new Bitcoin address generated from a pubkey held in the DeFiChain wallet.

spv_getrawtransaction TXID
Get raw transaction that relates to the provided TXID if found.

spv_listtransactions
Returns an array of all user related TXIDs in the wallet.

spv_sendtoaddress ADDRESS AMOUNT
Send Bitcoin.

spv_fundaddress (For function test framework)
Hidden command used in testing to fund the Bitcoin wallet via a dummy transaction.

### CODE REVIEW

File: src/spv/spv_rpc.cpp [hunk 1]
```diff
@@ -866,6 +848,213 @@ UniValue spv_setlastheight(const JSONRPCRequest& request)
...
+static UniValue spv_getnewaddress(const JSONRPCRequest& request)
+{
+    CWallet* const pwallet = GetWallet(request);
+
+    RPCHelpMan{"spv_getnewaddress",
+        "\nCreates and adds a Bitcoin address to the SPV wallet\n",
+        {
+        },
+        RPCResult{
+            "\"none\"                  Returns nothing\n"
```
Comment (ShengguangXiao) [inline src/spv/spv_rpc.cpp:888]: Here should returns the new address right?

File: src/spv/bitcoin/BRWallet.h [hunk 2]
```diff
@@ -56,11 +58,19 @@ inline static int BRUTXOEq(const void *utxo, const void *otherUtxo)
...
+// Convert SPV UInt256 to Bitcoin uint256
+uint256 to_uint256(UInt256 const & i);
+
 typedef struct BRWalletStruct BRWallet;
 
 // allocates and populates a BRWallet struct that must be freed by calling BRWalletFree()
 // forkId is 0 for bitcoin, 0x40 for b-cash
-BRWallet *BRWalletNew(BRTransaction *transactions[], size_t txCount, BRMasterPubKey mpk, int forkId);
+BRWallet *BRWalletNew(BRTransaction *transactions[], size_t txCount, BRMasterPubKey mpk, int forkId,
+                      std::set<UInt160, decltype(&UInt160Compare)>&& userAddresses);
```
Comment (ShengguangXiao) [inline src/spv/bitcoin/BRWallet.h:73]: Can define a type for `std::set<UInt160, decltype(&UInt160Compare)>`? I see it used in quite a lot places.

Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.h]: It's a Lambda comparator.

File: src/spv/spv_rpc.cpp (hunk 1 above)
Comment (Bushstar) [reply src/spv/spv_rpc.cpp:888]: I think so :)

File: src/spv/bitcoin/BRWallet.h (hunk 2 above)
Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.h]: I'll update with a struct.

Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.h]: @ShengguangXiao https://github.com/DeFiCh/ain/pull/282/commits/eaf75eff44be33b9139bf99a2a2f25dfc6108746

Comment (ShengguangXiao) [reply src/spv/bitcoin/BRWallet.h]: Actually what i mean is make std::set<UInt160, UInt160Compare> as SetOfAddress like below, but i think now is quite simple already. Thanks.
```
using SetOfAddress = std::set<UInt160, UInt160Compare>
```

Comment (Bushstar) [reply src/spv/bitcoin/BRWallet.h]: Oh right, I thought you did not like the somewhat cumbersome decltype(&UInt160Compare) as the comparator, I guess a using would have helped there. I often personally prefer to see what it is and not a using declaration in its place, with the decltype included a using declaration would have been better but I think it is okay now.

**Comment by ShengguangXiao** [comment]
@Bushstar I tested `spv_sendtoaddress` on testnet, if it still in syncing status, then this RPC call will hangup and timeout. After fully synced, it works ok.

**Comment by Bushstar** [comment]
@ShengguangXiao good spot. Will look into and remedy this on Monday.

**Comment by Bushstar** [comment]
@ShengguangXiao added a commit. Sending when SPV is not connected or less than the last days worth of blocks from the tip is not allowed.

**Review by ShengguangXiao — APPROVED** [review]

**Comment by ShengguangXiao** [comment]
> @ShengguangXiao added a commit. Sending when SPV is not connected or less than the last days worth of blocks from the tip is not allowed.

Now it works good.

**Review by monstrobishi — APPROVED** [review]

**Comment by monstrobishi** [comment]
@Bushstar how is this handling the change address?

**Pull request merged**